troposphere/haze/source/ptp_responder.cpp

```diff
@@ -0,0 +1,698 @@
+
+                /* Ensure we maintain a clean state on exit. */
+                ON_SCOPE_EXIT { m_fs.FileClose(std::addressof(file)); };
+
+                R_TRY(m_fs.FileGetSize(std::addressof(file), std::addressof(size)));
+            }
+
+            object_info.filename = std::strrchr(fileobj->GetName(), '/') + 1;
+            object_info.object_compressed_size = size;
+            object_info.parent_object = fileobj->GetParentId();
+
+            if (entry_type == FsDirEntryType_Dir) {
+                object_info.object_format = PtpObjectFormatCode_Association;
+                object_info.association_type = PtpAssociationType_GenericFolder;
+            } else {
+                object_info.object_format = PtpObjectFormatCode_Undefined;
+                object_info.association_type = PtpAssociationType_Undefined;
+            }
+        }
+
+        R_TRY(db.WriteVariableLengthData(m_request_header, [&] {
+            R_TRY(db.Add(object_info.storage_id));
+            R_TRY(db.Add(object_info.object_format));
+            R_TRY(db.Add(object_info.protection_status));
+            R_TRY(db.Add(object_info.object_compressed_size));
+            R_TRY(db.Add(object_info.thumb_format));
+            R_TRY(db.Add(object_info.thumb_compressed_size));
+            R_TRY(db.Add(object_info.thumb_width));
+            R_TRY(db.Add(object_info.thumb_height));
+            R_TRY(db.Add(object_info.image_width));
+            R_TRY(db.Add(object_info.image_height));
+            R_TRY(db.Add(object_info.image_depth));
+            R_TRY(db.Add(object_info.parent_object));
+            R_TRY(db.Add(object_info.association_type));
+            R_TRY(db.Add(object_info.association_desc));
+            R_TRY(db.Add(object_info.sequence_number));
+            R_TRY(db.AddString(object_info.filename));
+            R_TRY(db.AddString(object_info.capture_date));
+            R_TRY(db.AddString(object_info.modification_date));
+            R_TRY(db.AddString(object_info.keywords));
+
+            R_SUCCEED();
+        }));
+
+        R_RETURN(this->WriteResponse(PtpResponseCode_Ok));
+    }
+
+    Result PtpResponder::GetObject(PtpDataParser &dp) {
+        PtpDataBuilder db(s_bulk_write_buffer, std::addressof(m_usb_server));
+
+        /* Get the object ID the client requested. */
+        u32 object_id;
+        R_TRY(dp.Read(std::addressof(object_id)));
+        R_TRY(dp.Finalize());
+
+        /* Check if we know about the object. If we don't, it's an error. */
+        auto *fileobj = m_object_database.GetObject(object_id);
+        R_UNLESS(fileobj != nullptr, haze::ResultObjectNotFound());
+
+        /* Lock the object as a file. */
+        FsFile file;
+        R_TRY(m_fs.OpenFile(fileobj->GetName(), FsOpenMode_Read, std::addressof(file)));
+
+        /* Ensure we maintain a clean state on exit. */
+        ON_SCOPE_EXIT { m_fs.FileClose(std::addressof(file)); };
+
+        /* Get the file's size. */
+        s64 size = 0, offset = 0;
```

Comment:
I am confused why offset is declared up here. It's not used or modified until much later.